// @ngxs/devtools-plugin/esm5/src/devtools.module.js
import { NgModule, InjectionToken } from '@angular/core';
import { NgxsModule, NGXS_PLUGINS } from '@ngxs/store';
import { NGXS_DEVTOOLS_OPTIONS } from './symbols';
import { NgxsReduxDevtoolsPlugin } from './devtools.plugin';
export function devtoolsOptionsFactory(options) {
    return Object.assign({ name: 'NGXS' }, options);
}
export var USER_OPTIONS = new InjectionToken('USER_OPTIONS');
var NgxsReduxDevtoolsPluginModule = /** @class */ (function () {
    function NgxsReduxDevtoolsPluginModule() {
    }
    NgxsReduxDevtoolsPluginModule.forRoot = function (options) {
        return {
            ngModule: NgxsReduxDevtoolsPluginModule,
            providers: [
                {
                    provide: NGXS_PLUGINS,
                    useClass: NgxsReduxDevtoolsPlugin,
                    multi: true
                },
                {
                    provide: USER_OPTIONS,
                    useValue: options
                },
                {
                    provide: NGXS_DEVTOOLS_OPTIONS,
                    useFactory: devtoolsOptionsFactory,
                    deps: [USER_OPTIONS]
                }
            ]
        };
    };
    return NgxsReduxDevtoolsPluginModule;
}());
export { NgxsReduxDevtoolsPluginModule };
NgxsReduxDevtoolsPluginModule.decorators = [
    { type: NgModule, args: [{
                imports: [NgxsModule]
            },] },
];
